account-integrations/safe/src/SafeZkEmailRecoveryPlugin.sol

```diff
@@ -0,0 +1,133 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import {MockGroth16Verifier} from "./utils/MockGroth16Verifier.sol";
+import {ISafe} from "./utils/Safe4337Base.sol";
+
+interface ISafeECDSAPlugin {
+    function getOwner(address safe) external view returns (address);
+}
+
+struct ZkEmailRecoveryStorage {
+    bytes32 recoveryHash;
+}
+
+/**
+ * A safe plugin that recovers a safe ecdsa plugin owner via a zkp of an email.
+ * NOT FOR PRODUCTION USE
+ */
+contract SafeZkEmailRecoveryPlugin {
+    bytes32 immutable RECOVERY_HASH_DOMAIN;
+    string public constant DOMAIN_NAME = "RECOVERY_PLUGIN";
+    uint256 public constant DOMAIN_VERSION = 1;
+
+    mapping(address => ZkEmailRecoveryStorage) public zkEmailRecoveryStorage;
+
+    error INVALID_RECOVERY_HASH(
+        bytes32 recoveryHash,
+        bytes32 expectedRecoveryHash
+    );
+    error MODULE_NOT_ENABLED();
+    error INVALID_OWNER(address expectedOwner, address owner);
+    error INVALID_PROOF();
+
+    MockGroth16Verifier public immutable verifier;
+
+    constructor(address _verifier) {
+        verifier = MockGroth16Verifier(_verifier);
+        RECOVERY_HASH_DOMAIN = keccak256(
+            abi.encodePacked(
+                DOMAIN_NAME,
+                DOMAIN_VERSION,
+                block.chainid,
+                address(this)
+            )
+        );
```

Comment:
Done. Although since we are now using the `recoveryHash` as a public signal into the verifier to hide the email and salt, we are no longer using `RECOVERY_HASH_DOMAIN` in this contract directly apart from assigning it in the constructor. Couldn't decide if passing it into the verifier was the right choice. What did you envision here?

Reply:
This code is still going to have issues if the `chainid` of the underlying network changes or after a fork. However, we may not even want/need to have a specific contract deployment domain for this initial recovery process. Can discuss more offline, and leave as is for now.